### PR TITLE
refactor: http

### LINKED ISSUE #2370: create a HTTP client per thread
Currently Tailcall's HTTP client is created once and shared across all its threads, during times of high-contention it can create performance issues. We propose that an http-client is created per thread lazily with each incoming request. 

**Technical Requirements**
- All existing tests should pass
- There should be no code duplication
- It should overall improve performance (Check via the nginx benchmark on CI)

### PR DESCRIPTION
**Pull request by Shylock-Hg**
**Summary:**  
Create http for each thread 

**Issue Reference(s):**  
Close #2370 
/claim #2370 

**Build & Testing:**

- [x] I ran `cargo test` successfully.
- [x] I have run `./lint.sh --mode=fix` to fix all linting issues raised by `./lint.sh --mode=check`.

**Checklist:**

- [ ] I have added relevant unit & integration tests.
- [ ] I have updated the [documentation] accordingly.
- [x] I have performed a self-review of my code.
- [x] PR follows the naming convention of `<type>(<optional scope>): <title>`

[documentation]: https://github.com/tailcallhq/tailcallhq.github.io/tree/develop/docs

### CODE REVIEW

File: src/cli/runtime/http.rs [hunk 1]
```diff
@@ -182,8 +183,41 @@ impl HttpIO for NativeHttp {
         )
         .await?)
     }
+
+    fn cl(&self) -> Box<dyn HttpIO> {
+        Box::new(self.clone())
```
Comment (tusharmath) [inline src/cli/runtime/http.rs:188]: Cloning isn't going to help, ideally we want to create separate connection pools and caches for each http-client.

File: src/cli/runtime/http.rs [hunk 2]
```diff
@@ -182,8 +183,41 @@ impl HttpIO for NativeHttp {
...
 }
 
+#[async_trait::async_trait]
+impl HttpIO for ConcurrentHttp {
+    async fn execute(&self, request: reqwest::Request) -> Result<Response<Bytes>> {
+        let thread_id = std::thread::current().id();
+        if let Some(http) = self.map.get(&thread_id) {
+            http.value().execute(request).await
+        } else {
+            let new_http = self.http.cl();
```
Comment (tusharmath) [inline src/cli/runtime/http.rs:210]: Should call `init_http` instead.

File: src/cli/runtime/http.rs [hunk 3]
```diff
@@ -182,8 +183,41 @@ impl HttpIO for NativeHttp {
...
     }
+
+    fn cl(&self) -> Box<dyn HttpIO> {
+        Box::new(self.clone())
+    }
+}
+
+pub struct ConcurrentHttp {
+    http: Box<dyn HttpIO>,
+    map: DashMap<std::thread::ThreadId, Box<dyn HttpIO>>,
```
Comment (tusharmath) [inline src/cli/runtime/http.rs:194]: instead of map we can use a thread local.

File: src/cli/runtime/http.rs (hunk 1 above)
Comment (Shylock-Hg) [reply src/cli/runtime/http.rs:188]: I have a question about this. According to `https://docs.rs/reqwest/0.12.5/reqwest/struct.Client.html`, `reqwest::Client` has a connection pool in it already. Do we need more work?

Comment (tusharmath) [reply src/cli/runtime/http.rs:188]: Nothing special needs to be done for pooling.

**Comment by tusharmath** [comment]
Moving to draft to reduce noise and improve CI efficiency. Once you are ready just mark it as "ready to review". Feel free to give a shoutout on the #contributors channel on Discord if you want immediate attention.

Let me know once the comments are fixed. Thanks!

File: src/cli/runtime/http.rs (hunk 2 above)
Comment (Shylock-Hg) [reply src/cli/runtime/http.rs]: Ok

File: src/cli/runtime/http.rs (hunk 3 above)
Comment (Shylock-Hg) [reply src/cli/runtime/http.rs]: Ok

**Comment by Shylock-Hg** [comment]
@tusharmath I think it's ok now. But the CI of benchmark went wrong.

**Comment by tusharmath** [comment]
@Shylock-Hg The PR isn't doing whats expected:
1. Benchmarks are failing, meaning the server doesn't work correctly.
2. There is use of unsafe code such as `panic`.
3. You are still creating a single client and then cloning it in an arch which is exactly what we don't want. 

For these reasons, I am closing this PR for now. Feel free to create a new PR and ask questions on discord if the requirement isn't clear, always happy to help!

**Comment by Shylock-Hg** [comment]
> @Shylock-Hg The PR isn't doing whats expected:
> 
> 1. Benchmarks are failing, meaning the server doesn't work correctly.
> 2. There is use of unsafe code such as `panic`.
> 3. You are still creating a single client and then cloning it in an arch which is exactly what we don't want.
> 
> For these reasons, I am closing this PR for now. Feel free to create a new PR and ask questions on discord if the requirement isn't clear, always happy to help!

Ok, I misunderstand it

**Pull request closed without merging**